src/telemetry/metrics/prom.rs

```diff
@@ -0,0 +1,126 @@
+use std::fmt;
+use std::marker::PhantomData;
+
+/// Writes a block of metrics in prometheus-formatted output.
+pub trait FmtPrometheus {
+    fn fmt_prometheus(&self, f: &mut fmt::Formatter) -> fmt::Result;
+}
+
+/// Writes a series of key-quoted-val pairs for use as prometheus labels.
+pub trait FmtLabels {
+    fn fmt_labels(&self, f: &mut fmt::Formatter) -> fmt::Result;
+}
+
+/// Writes a metric in prometheus-formatted output.
+///
+/// This trait is implemented by `Counter`, `Gauge`, and `Histogram` to account for the
+/// differences in formatting each type of metric. Specifically, `Histogram` formats a
+/// counter for each bucket, as well as a count and total sum.
+pub trait FmtMetric {
+    /// The metric's `TYPE` in help messages.
+    const KIND: &'static str;
+
+    /// Writes a metric with the given name and no labels.
+    fn fmt_metric<N: fmt::Display>(&self, f: &mut fmt::Formatter, name: N) -> fmt::Result;
+
+    /// Writes a metric with the given name and labels.
+    fn fmt_metric_labeled<N, L>(&self, f: &mut fmt::Formatter, name: N, labels: L) -> fmt::Result
+    where
+        N: fmt::Display,
+        L: FmtLabels;
+}
+
+/// Describes a metric statically.
+///
+/// Formats help messages and metric values for prometheus output.
+pub struct Metric<'a, M: FmtMetric> {
+    pub name: &'a str,
+    pub help: &'a str,
+    pub _p: PhantomData<M>,
+}
+
+// ===== impl Metric =====
+
+impl<'a, M: FmtMetric> Metric<'a, M> {
+    /// Formats help messages for this metric.
+    pub fn fmt_help(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        writeln!(f, "# HELP {} {}", self.name, self.help)?;
+        writeln!(f, "# TYPE {} {}", self.name, M::KIND)?;
+        Ok(())
+    }
+
+    /// Formats a single metric without labels.
+    pub fn fmt_metric(&self, f: &mut fmt::Formatter, metric: M) -> fmt::Result {
+        metric.fmt_metric(f, self.name)
+    }
+
+    /// Formats a single metric across labeled scopes.
+    pub fn fmt_scopes<'s, L, S: 's, I, F>(
+        &self,
+        f: &mut fmt::Formatter,
+        scopes: I,
+        to_metric: F
+    ) -> fmt::Result
+    where
+        L: FmtLabels,
+        I: IntoIterator<Item = (L, &'s S)>,
+        F: Fn(&S) -> &M
+    {
+        for (labels, scope) in scopes {
+            to_metric(scope).fmt_metric_labeled(f, self.name, labels)?;
+        }
+
+        Ok(())
+    }
+}
+
+// ===== impl FmtLabels =====
+
+impl<'a, A: FmtLabels + 'a> FmtLabels for &'a A {
+    fn fmt_labels(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        (*self).fmt_labels(f)
+    }
+}
+
+impl<A: FmtLabels, B: FmtLabels> FmtLabels for (A, B) {
```

Comment:
The helper methods on tuples are cute. I wonder if it would be simpler to just define one on `IntoIterator<Item=&'a FmtLabels>`, though? Then, the case of `Option`al labels could be handled with `filter_map`.

Reply:
we need to know that there is at least one valid label, otherwise all of the comma insertion breaks.  This is why we dont provide impls for `Option<A>` or `(Option<A>, Option<B>)`...

are you suggesting that we support `(A, IntoIterator<Item=FmtLabels>)` instead of the Option? would that work?

Reply:
> we need to know that there is at least one valid label, otherwise all of the comma insertion breaks. 

I think we can do something like this:
```rust 
let mut iter = i.into_iter() // iterator of Option<FmtLabels>
    .filter_map(|i| i); // filter out any nones.

if let Some(label) = iter.next() {
    write!(f, "{}", label)?;

    for label in iter {
        write!(f, ",{}", label)?;
    }
}
```

regardless, this seems fine as-is!

Reply:
I think this has two problems:

1. `FmtLabels` must not represent an empty set of labels. Either that, or the `FmtLabels` type would need to expose a `is_empty()`. This is important for the purposes of composition.  If I have an iterator of `FmtLabels` that i want to combine, how do i know if any one of those items is itself empty?
2. AIUI, we would almost always need to use `Box<FmtLabels>`, because we are trying to handle diverse types.